File: Milla/Stack/MinStack.java
package Milla.Stack;

import java.util.Stack;

public class MinStack {
    /*
    *
    *
    * */
    static Stack<Integer> stack;
    static int min;
    public static void main(String[] args) {
        MinStack minStack = new MinStack();
        minStack.push(1);
        minStack.push(2);
        minStack.push(0);
        System.out.println(minStack.getMin()); // return 0
        minStack.pop();
        System.out.println(minStack.top());    // return 2
        System.out.println(minStack.getMin());
    }

    public MinStack() {
        min = Integer.MAX_VALUE;
        stack = new Stack<>();
    }

    public void push(int val) {
        if(val <= min){
            stack.push(min);
            min = val;
        }
        stack.push(val);
    }

    public void pop() {
        if(stack.pop() == min){
            int sndMin = stack.pop();
            min = sndMin;
        }
    }

    public int top() {
        return stack.peek();
    }

    public int getMin() {
        return min;
    }
}

File: Joy/Stack DS/min-stack.py
'''
❓ Design a stack that supports push, pop, top and get minimum element in O(1)

Clarifications
1. Each method should be implemented in O(1)
2. If pop, getMin, top method is called on empty stack, do I return None? -> Yes

Brainstorm
- brute force: If getMin is called, go through the stack and find the minimum value O(N)
- optimal approach: When push the element, use tuple to store the (value, minimum up to that point) O(1)

=> TC: O(1), SC: O(N)
'''

class MinStack:
    def __init__(self):
        self.stack = []

    def push(self, val):
        if not self.stack:
            self.stack.append((val, val))
        else:
            self.stack.append((val, min(val, self.stack[-1][1])))

    def pop(self):
        self.stack.pop()
    
    def top(self):
        return self.stack[-1][0]

    def getMin(self):
        return self.stack[-1][1]

